viewing_party/party.py

```diff
@@ -1,23 +1,145 @@
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    sum1=[]
+    av_rait=0
+    user_data = user_data["watched"]
+    for i in user_data:
+       
+        sum1.append(i['rating'])
+        av_rait = sum(sum1)/len(sum1)
+
+    return av_rait
+        
+    
+def get_most_watched_genre(user_data):
+
+    user_data = user_data['watched']
+    for i in user_data:
+        for b in user_data:
+            if i['genre']== b['genre']:
+                return i['genre']
+            else:
+                return None    
+
+
+
+
 
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+    unique_movie=[]
+    movies = []
+    for friend in user_data['friends']:
+        for movie in friend['watched']:
+            movies.append(movie)
+    for i in user_data['watched']:
+        if i not in movies:
+            unique_movie.append(i)
+                        
+
+    return unique_movie            
+
+def get_friends_unique_watched(user_data):
+    unique_movie=[]
+    friend_unique = []
+    for friend in user_data['friends']:
+        for movie in friend['watched']:
+            if movie not in unique_movie:
+                unique_movie.append(movie)
+    for i in unique_movie:
+        if i not in user_data['watched']:
+            friend_unique.append(i)
+    return friend_unique
+
+                        
 
-        
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
-
+def get_available_recs(user_data):
+    unique_movie=[]
+    friend_unique = []
+    for friend in user_data['friends']:
+        for movie in friend['watched']:
+            if movie not in unique_movie:
+                unique_movie.append(movie)
+    for i in unique_movie:
+        if i not in user_data['watched'] and i['host'] in user_data['subscriptions']:
+            friend_unique.append(i)
+    return friend_unique
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+    fr_genre = []
+    un_genre = str
+    for genre in user_data['watched']:
+        fr_genre.append(genre['genre'])
+        if len(user_data['watched'])==0:
+            un_genre= None
+        else:    
+            un_genre = max(set(fr_genre), key = fr_genre.count)    
+
+    rec_movie=[]
+    friend_movies = []
+    for friend in user_data['friends']:
+        for movie in friend['watched']:
+            if movie not in friend_movies:
+                friend_movies.append(movie)
+    for i in friend_movies:
+        if i not in user_data['watched'] and i['genre'] == un_genre:
+            rec_movie.append(i) 
+    return rec_movie        

```

Comment:
I noticed you were grabbing the friends movies using an iteration, you could also use previous functions to grab that data for you. You could also grab the unique genre doing something like this:
```
def get_new_rec_by_genre(user_data):
    genre_rec = []
    users_most_watched_genre = get_most_watched_genre(user_data)
    friends_watched = user_data["friends"]

    for friend in friends_watched:
        for movie in friend["watched"]:
            if movie["genre"] == users_most_watched_genre and movie not in user_data["watched"] and movie not in genre_rec:
                genre_rec.append(movie)

    return genre_rec
```